crates/loro-internal/src/state.rs

```diff
@@ -121,27 +106,109 @@ pub(crate) trait ContainerState: Clone {
     fn import_from_snapshot_ops(&mut self, ctx: StateSnapshotDecodeContext);
 }
 
+impl<T: ContainerState> ContainerState for Box<T> {
+    fn container_idx(&self) -> ContainerIdx {
+        self.as_ref().container_idx()
+    }
+
+    fn estimate_size(&self) -> usize {
+        self.as_ref().estimate_size()
+    }
+
+    fn is_state_empty(&self) -> bool {
+        self.as_ref().is_state_empty()
+    }
+
+    fn apply_diff_and_convert(
+        &mut self,
+        diff: InternalDiff,
+        arena: &SharedArena,
+        txn: &Weak<Mutex<Option<Transaction>>>,
+        state: &Weak<Mutex<DocState>>,
+    ) -> Diff {
+        self.as_mut()
+            .apply_diff_and_convert(diff, arena, txn, state)
+    }
+
+    fn apply_diff(
+        &mut self,
+        diff: InternalDiff,
+        arena: &SharedArena,
+        txn: &Weak<Mutex<Option<Transaction>>>,
+        state: &Weak<Mutex<DocState>>,
+    ) {
+        self.as_mut().apply_diff(diff, arena, txn, state)
+    }
+
+    fn apply_op(&mut self, raw_op: &RawOp, op: &Op, arena: &SharedArena) -> LoroResult<()> {
+        self.as_mut().apply_op(raw_op, op, arena)
+    }
+
+    #[doc = r" Convert a state to a diff, such that an empty state will be transformed into the same as this state when it's applied."]
+    fn to_diff(
+        &mut self,
+        arena: &SharedArena,
+        txn: &Weak<Mutex<Option<Transaction>>>,
+        state: &Weak<Mutex<DocState>>,
+    ) -> Diff {
+        self.as_mut().to_diff(arena, txn, state)
+    }
+
+    fn get_value(&mut self) -> LoroValue {
+        self.as_mut().get_value()
+    }
+
+    #[doc = r" Get the index of the child container"]
+    #[allow(unused)]
+    fn get_child_index(&self, id: &ContainerID) -> Option<Index> {
+        self.as_ref().get_child_index(id)
+    }
+
+    #[allow(unused)]
+    fn get_child_containers(&self) -> Vec<ContainerID> {
+        self.as_ref().get_child_containers()
+    }
+
+    #[doc = r" Encode the ops and the blob that can be used to restore the state to the current state."]
+    #[doc = r""]
+    #[doc = r" State will use the provided encoder to encode the ops and export a blob."]
+    #[doc = r" The ops should be encoded into the snapshot as well as the blob."]
+    #[doc = r" The users then can use the ops and the blob to restore the state to the current state."]
+    fn encode_snapshot(&self, encoder: StateSnapshotEncoder) -> Vec<u8> {
+        self.as_ref().encode_snapshot(encoder)
+    }
+
+    #[doc = r" Restore the state to the state represented by the ops and the blob that exported by `get_snapshot_ops`"]
+    fn import_from_snapshot_ops(&mut self, ctx: StateSnapshotDecodeContext) {
+        self.as_mut().import_from_snapshot_ops(ctx)
+    }
+}
+
 #[allow(clippy::enum_variant_names)]
 #[enum_dispatch(ContainerState)]
 #[derive(EnumAsInner, Clone, Debug)]
 pub enum State {
-    ListState,
-    MapState,
-    RichtextState,
-    TreeState,
+    ListState(Box<ListState>),
```

Comment:
why `Box`

Reply:
It can reduce the memory waste caused by the different sizes between the variants.

Reply:
But it is redundant if every state has a similar size. I can have a try.

Reply:
I think it's better to Box them. Because each one of them has very different size:

```
Size of MapState = 40
Size of ListState = 128
Size of TextState = 168
Size of TreeState = 72
```

And I found it hard to make all of them around 40 bytes

Reply:
👌🏻